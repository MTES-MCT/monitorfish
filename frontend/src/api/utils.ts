import { BackendApi } from '@api/BackendApi.types'
import { HttpStatusCode } from '@api/constants'
import { isString } from 'lodash-es'

import type { SafeParseReturnType } from 'zod'

export function isUnauthorizedOrForbidden(httpStatus: number | string | undefined) {
  if (!httpStatus || isString(httpStatus)) {
    return false
  }

  return [HttpStatusCode.FORBIDDEN, HttpStatusCode.UNAUTHORIZED].includes(httpStatus)
}

export function valueOrUndefinedIfNotFoundOrThrow<Type>(
  result: SafeParseReturnType<any, any>,
  response: BackendApi.ResponseBodyError | Type
): Type | undefined {
  if (!result.success) {
    if ((response as BackendApi.ResponseBodyError).code === BackendApi.ErrorCode.NOT_FOUND_BUT_OK) {
      return undefined
    }

    throw result.error
  }

  return response as Type
}
