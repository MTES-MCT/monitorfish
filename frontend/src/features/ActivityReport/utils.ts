import { range, sortBy } from 'lodash'

import { JDP, UNTARGETED_SPECIES_CODE } from './constants'
import { MissionAction } from '../Mission/missionAction.types'

import type { ActivityReportWithId } from './types'
import type { DownloadAsCsvMap } from '@utils/downloadAsCsv'

export function formatDMDCoordinateForActivityReport(coordinate: string | undefined): string {
  const hemisphere = coordinate?.at(-1)
  if (!hemisphere) {
    return ''
  }

  const nextCoordinate = coordinate?.replace(/°/g, '')?.replace(/′/g, '')?.replace(` ${hemisphere}`, '')

  return `${hemisphere}${nextCoordinate}`
}

/**
 * Adds 'SPECIES' related columns, 'INFR' related columns and 'COMMENT' column to JDP_CSV_MAP_BASE
 */
export function getJDPCsvMap(baseCsvMap: DownloadAsCsvMap<ActivityReportWithId>, jdp: JDP) {
  const numberOfSpeciesColumns = 10
  const numberOfInfractionColumns = 10

  range(numberOfSpeciesColumns).forEach(index => {
    const count = index + 1

    // eslint-disable-next-line no-param-reassign
    baseCsvMap[`species${count}`] = {
      label: `SPECIES${count}`,
      transform: activity => {
        const speciesOnboard = activity.action.speciesOnboard[count - 1]
        if (!speciesOnboard?.speciesCode) {
          return ''
        }

        return speciesOnboard.speciesCode
      }
    }

    // eslint-disable-next-line no-param-reassign
    baseCsvMap[`weight${count}`] = {
      label: `WEIGHT${count}`,
      transform: activity => {
        const speciesOnboard = activity.action.speciesOnboard[count - 1]

        return speciesOnboard?.controlledWeight ?? speciesOnboard?.declaredWeight ?? ''
      }
    }

    // eslint-disable-next-line no-param-reassign
    baseCsvMap[`nbFish${count}`] = {
      label: `NB_IND${count}`,
      transform: activity => {
        const speciesOnboard = activity.action.speciesOnboard[count - 1]

        return speciesOnboard?.nbFish ?? ''
      }
    }
  })

  range(numberOfInfractionColumns).forEach(index => {
    const count = index + 1

    // eslint-disable-next-line no-param-reassign
    baseCsvMap[`infractionClass${count}`] = jdp === JDP.WESTERN_WATERS ? `INFR${count}_CLASS` : `INFR_CLASS${count}`

    // eslint-disable-next-line no-param-reassign
    baseCsvMap[`infractionCode${count}`] = {
      label: jdp === JDP.WESTERN_WATERS ? `INFR${count}_CODE` : `INFR_CODE${count}`,
      transform: activity => {
        const allNatinfs = getInfractionsKeys(activity.action, 'natinf')

        return allNatinfs[count - 1] ?? ''
      }
    }

    // eslint-disable-next-line no-param-reassign
    baseCsvMap[`infractionDescription${count}`] = {
      label: jdp === JDP.WESTERN_WATERS ? `INFR${count}_DESCRIPTION` : `INFR_REMARK${count}`,
      transform: activity => {
        const allComments = getInfractionsKeys(activity.action, 'comments')

        return allComments[count - 1] ?? ''
      }
    }
  })

  // eslint-disable-next-line no-param-reassign
  baseCsvMap['action.otherComments'] = `COMMENT`

  return baseCsvMap
}

export function getSpeciesOnboardWithUntargetedSpeciesGrouped(
  speciesOnboard: MissionAction.SpeciesControl[],
  jdpSpecies: string[]
): MissionAction.SpeciesControl[] {
  const otherSpeciesSummedWeight = speciesOnboard
    .filter(species => !jdpSpecies.includes(species.speciesCode))
    .map(species => species.controlledWeight ?? species.declaredWeight)
    .filter((species): species is number => species !== undefined)
    .reduce((accumulator, currentValue) => accumulator + currentValue, 0)

  const otherSpecy = {
    controlledWeight: undefined,
    declaredWeight: otherSpeciesSummedWeight,
    nbFish: undefined,
    speciesCode: UNTARGETED_SPECIES_CODE,
    underSized: undefined
  }

  const speciesOnboardWithoutOtherSpecies = speciesOnboard.filter(species => jdpSpecies.includes(species.speciesCode))

  const groupedSpeciesOnboard =
    otherSpeciesSummedWeight > 0
      ? speciesOnboardWithoutOtherSpecies.concat(otherSpecy)
      : speciesOnboardWithoutOtherSpecies

  return sortBy(groupedSpeciesOnboard, ({ declaredWeight }) => declaredWeight).reverse()
}

function getInfractionsKeys(action: MissionAction.MissionAction, key: string): string[] {
  return ([] as string[]).concat.apply(
    [],
    [
      getInfractionsWithRecordKey(action.gearInfractions, key),
      getInfractionsWithRecordKey(action.speciesInfractions, key),
      getInfractionsWithRecordKey(action.logbookInfractions, key),
      getInfractionsWithRecordKey(action.otherInfractions, key)
    ]
  )
}

function getInfractionsWithRecordKey(infractions: MissionAction.Infraction[], key: string): string[] {
  return infractions
    .filter(infraction => infraction.infractionType === MissionAction.InfractionType.WITH_RECORD)
    .map(infraction => infraction[key])
}

export function getPatrolType(activity: ActivityReportWithId): string {
  if (activity.action.actionType === MissionAction.MissionActionType.SEA_CONTROL) {
    return 'S'
  }

  if (activity.action.actionType === MissionAction.MissionActionType.LAND_CONTROL) {
    return 'L'
  }

  if (activity.action.actionType === MissionAction.MissionActionType.AIR_CONTROL) {
    return 'A'
  }

  return ''
}
