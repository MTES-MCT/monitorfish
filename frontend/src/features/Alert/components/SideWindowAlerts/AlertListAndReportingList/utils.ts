import { COMMON_ALERT_TYPE_OPTION } from '../../../constants'

import type { PendingAlertValueType } from '../../../types'

export const getAlertNameFromType = (type: PendingAlertValueType | 'PNO_LAN_WEIGHT_TOLERANCE_ALERT') =>
  COMMON_ALERT_TYPE_OPTION[type] ? COMMON_ALERT_TYPE_OPTION[type].name : 'Alerte inconnue'
