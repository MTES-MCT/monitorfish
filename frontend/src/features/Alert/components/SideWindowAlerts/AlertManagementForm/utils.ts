import { EU_COUNTRY_CODES } from '@features/Alert/components/SideWindowAlerts/AlertManagementForm/constants'
import { AdministrativeAreaType, AdministrativeAreaTypeLabel } from '@features/Alert/constants'
import Countries, { getAlpha3Code, getNames } from 'i18n-iso-countries'
import COUNTRIES_FR from 'i18n-iso-countries/langs/fr.json'
import { flatMap, sortBy } from 'lodash-es'

import type { TreeOption } from '@mtes-mct/monitor-ui'
import type { TreeBranchOption, TreeLeafOption } from '@mtes-mct/monitor-ui/types/definitions'

Countries.registerLocale(COUNTRIES_FR)

const LABEL_TO_AREA_TYPE = {
  [AdministrativeAreaTypeLabel.EEZ_AREA]: AdministrativeAreaType.EEZ_AREA,
  [AdministrativeAreaTypeLabel.FAO_AREA]: AdministrativeAreaType.FAO_AREA
}

/**
 * Extracts the appropriate zone code based on area type
 */
const getZoneCode = (zoneLabel: string, areaType: AdministrativeAreaType): string => {
  if (areaType === AdministrativeAreaType.EEZ_AREA) {
    // For EEZ areas, convert country name to ISO Alpha3 code
    const alpha3Code = getAlpha3Code(zoneLabel, 'en')

    return alpha3Code ?? zoneLabel
  }

  // For FAO areas, use the label as-is (e.g., "27.1.0")
  return zoneLabel
}

/**
 * Maps zone groups to flat array with metadata (ID, area type, and zone code)
 * @param zoneGroups - Tree options containing zone groups and their children
 * @returns Array of zones with their metadata
 */
export const mapZonesWithMetadata = (zoneGroups: TreeOption[] | undefined) =>
  flatMap(zoneGroups, zoneGroup => {
    const areaType = LABEL_TO_AREA_TYPE[zoneGroup.label]

    return areaType && zoneGroup.children
      ? zoneGroup.children.map(zone => {
          const zoneCode = getZoneCode(zone.label, areaType)

          return {
            areaType,
            code: zoneCode,
            id: zone.value
          }
        })
      : []
  })

/**
 * Builds TreeOption[] with EU and Non-EU countries using i18n-iso-countries
 * @param locale - Language locale for country names (default: 'fr')
 * @returns TreeOption array with EU and Non-EU country groups
 */
export const buildCountriesAsTreeOptions = (locale: string = 'fr'): TreeOption[] => {
  const allCountries = getNames(locale)
  const euCountries: TreeOption[] = []
  const nonEuCountries: TreeOption[] = []

  Object.entries(allCountries).forEach(([alpha2Code, countryName]) => {
    const countryOption: TreeLeafOption = {
      label: countryName,
      value: alpha2Code
    }

    if (EU_COUNTRY_CODES.includes(alpha2Code)) {
      euCountries.push(countryOption)
    } else {
      nonEuCountries.push(countryOption)
    }
  })

  const sortedEuCountries = sortBy(euCountries, 'label')
  const sortedNonEuCountries = sortBy(nonEuCountries, 'label')

  return [
    {
      children: sortedEuCountries,
      label: 'Navires UE',
      value: 'eu'
    } as TreeBranchOption,
    {
      children: sortedNonEuCountries,
      label: 'Navires tiers',
      value: 'non-eu'
    } as TreeBranchOption
  ]
}
