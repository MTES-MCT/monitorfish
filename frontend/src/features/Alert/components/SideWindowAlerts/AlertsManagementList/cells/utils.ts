import type { AlertSpecification } from '@features/Alert/types'
import type { Column } from '@tanstack/react-table'
import type { CSSProperties } from 'react'

export function getExpandableRowCellCustomStyle(column: Column<AlertSpecification, any>): CSSProperties {
  const defaultStyle = {
    maxWidth: column.getSize()
  }

  switch (column.id) {
    case 'actions':
      return { ...defaultStyle, verticalAlign: 'bottom' }
    case 'validityPeriod':
      return { ...defaultStyle, verticalAlign: 'middle' }
    case 'isActivated':
      return { ...defaultStyle, paddingLeft: 11, verticalAlign: 'middle' }

    default:
      return defaultStyle
  }
}

export function getAlertCriteriaSummary(alertSpecification: AlertSpecification): string {
  let summary: string[] = []

  if (alertSpecification.vesselIds.length > 0) {
    summary = summary.concat('Navires')
  }
  if (alertSpecification.species.length > 0) {
    summary = summary.concat('Espèces à bord')
  }
  if (alertSpecification.speciesCatchAreas.length > 0) {
    summary = summary.concat('Zones de capture (FAR)')
  }
  if (alertSpecification.flagStatesIso2.length > 0) {
    summary = summary.concat('Nationalités')
  }
  if (alertSpecification.districtCodes.length > 0) {
    summary = summary.concat('Quartiers')
  }
  if (alertSpecification.producerOrganizations.length > 0) {
    summary = summary.concat('OPs')
  }
  if (alertSpecification.administrativeAreas.length > 0 || alertSpecification.regulatoryAreas.length > 0) {
    summary = summary.concat('Zones (VMS)')
  }
  if (alertSpecification.gears.length > 0) {
    summary = summary.concat('Engins')
  }

  return summary.join(', ')
}
