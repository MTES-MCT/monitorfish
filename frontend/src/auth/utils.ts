import { isUnauthorizedOrForbidden } from '@api/utils'
import { trackEvent } from '@hooks/useTracking'

import { ROUTER_PATHS } from '../paths'

import type { CustomResponseError } from '@api/types'

/**
 * Redirect to Login page if any HTTP request in Unauthorized
 */
export function redirectToLoginIfUnauthorized(error: CustomResponseError, email?: string | undefined) {
  if (!error.path.includes(ROUTER_PATHS.backendForFrontend) || !isUnauthorizedOrForbidden(error.status)) {
    return
  }

  if (!window.location.pathname.includes(ROUTER_PATHS.login)) {
    trackEvent({
      action: 'Redirection vers la page de login apr√®s une erreur API 401',
      category: 'REDIRECTION',
      name: email ?? ''
    })

    // We don't use `router.navigate()` to avoid circular dependency issues
    window.location.href = ROUTER_PATHS.login
  }
}
