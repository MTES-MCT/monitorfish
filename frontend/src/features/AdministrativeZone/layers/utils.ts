import { OPENLAYERS_PROJECTION, WSG84_PROJECTION } from '@features/Map/constants'
import GeoJSON from 'ol/format/GeoJSON'
import VectorImageLayer from 'ol/layer/VectorImage'
import { all, bbox as bboxStrategy } from 'ol/loadingstrategy'
import VectorSource from 'ol/source/Vector'

import { getAdministrativeLayerStyle } from './styles/administrativeLayer.style'
import { getAdministrativeZoneFromAPI } from '../../../api/geoserver'

import type Feature from 'ol/Feature'
import type Geometry from 'ol/geom/Geometry'

export const getVectorOLLayer = (
  type: string,
  zone: string | undefined,
  isBackoffice: boolean
): VectorImageLayer<Feature<Geometry>> => {
  const layer = new VectorImageLayer({
    className: 'administrative',
    declutter: true,
    source: getVectorSource(type, zone, isBackoffice),
    style: feature => [getAdministrativeLayerStyle(type)(feature as Feature)]
  })
  // @ts-ignore
  layer.name = getLayerNameFromTypeAndZone(type, zone)

  return layer
}

const getVectorSource = (type: string, zone: string | undefined, isBackoffice: boolean): VectorSource => {
  if (zone) {
    return buildWholeVectorSource(type, zone, isBackoffice)
  }

  return buildBBOXVectorSource(type, zone, isBackoffice)
}

function buildWholeVectorSource(
  type: string,
  zone: string | undefined,
  isBackoffice: boolean
): VectorSource<Feature<Geometry>> {
  const vectorSource = new VectorSource({
    format: new GeoJSON({
      dataProjection: WSG84_PROJECTION,
      featureProjection: OPENLAYERS_PROJECTION
    }),
    strategy: all
  })

  getAdministrativeZoneFromAPI(type, undefined, zone, isBackoffice).then(administrativeZoneFeature => {
    const features = vectorSource.getFormat()?.readFeatures(administrativeZoneFeature)
    if (!features) {
      return
    }

    vectorSource.addFeatures(features as Feature[])
  })

  return vectorSource
}

function buildBBOXVectorSource(
  type: string,
  zone: string | undefined,
  isBackoffice: boolean
): VectorSource<Feature<Geometry>> {
  const vectorSource = new VectorSource({
    format: new GeoJSON({
      dataProjection: WSG84_PROJECTION,
      featureProjection: OPENLAYERS_PROJECTION
    }),
    loader: extent => {
      getAdministrativeZoneFromAPI(type, extent, zone, isBackoffice)
        .then(administrativeZone => {
          vectorSource.clear(true)
          const features = vectorSource.getFormat()?.readFeatures(administrativeZone)
          if (!features) {
            return
          }

          vectorSource.addFeatures(features as Feature[])
        })
        .catch(() => {
          vectorSource.removeLoadedExtent(extent)
        })
    },
    strategy: bboxStrategy
  })

  return vectorSource
}

export function getLayerNameFromTypeAndZone(type: string, zone: string | undefined): string {
  if (!zone) {
    return type
  }

  return `${type}:${zone}`
}
