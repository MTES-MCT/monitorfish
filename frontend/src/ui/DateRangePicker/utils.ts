// TODO Use `date-fns` instead of `dayjs`.

import { dayjs } from '../../utils/dayjs'

import type { Option } from '../../types'
import type { DateTuple, TimeTuple } from './types'

export function formatNumberAsDoubleDigit(number: number): string {
  return String(number).padStart(2, '0')
}

export function getDateFromDateAndTimeTuple(dateTuple: DateTuple, timeTuple: TimeTuple, isEnd: boolean = false): Date {
  const [year, month, day] = dateTuple
  const [hour, minute] = timeTuple

  const rawDateAsDayjs = dayjs()
    .year(year)
    .month(month - 1)
    .date(day)
    .hour(hour)
    .minute(minute)

  return isEnd
    ? rawDateAsDayjs
        .endOf('minute')
        // TODO For some reason the API can't handle miliseconds in date.
        // That's why we set it to 0 (instead of 999)
        .millisecond(0)
        .toDate()
    : rawDateAsDayjs.startOf('minute').toDate()
}

export function getDateTupleFromDate(date?: Date): DateTuple | undefined {
  if (!date) {
    return undefined
  }

  return [date.getFullYear(), date.getMonth() + 1, date.getDate()]
}

/**
 * Generate a list of ranged time options.
 *
 * @example
 * ```
 * (minutesRange = 30) => ([
 *   { label: '00:00', value: [0, 0] },
 *   { label: '00:30', value: [0, 30] },
 *   { label: '01:00', value: [1, 0] },
 *   { label: '01:30', value: [1, 30] },
 *   { label: '02:00', value: [2, 0] },
 *   ...
 * ])
 * ```
 */
export const getRangedTimeOptions = (minutesRange: number): Option<TimeTuple>[] => {
  const perHourOptionsLength = 60 / minutesRange
  const totalOptionsLength = 24 * perHourOptionsLength

  return new Array(totalOptionsLength).fill(undefined).map((_, index) => {
    const hour = Math.floor(index / perHourOptionsLength)
    const minute = minutesRange * (index % perHourOptionsLength)
    const label = `${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}`
    const value: TimeTuple = [hour, minute]

    return {
      label,
      value,
    }
  })
}

export function getTimeTupleFromDate(date?: Date): TimeTuple | undefined {
  if (!date) {
    return undefined
  }

  return [date.getHours(), date.getMinutes()]
}
