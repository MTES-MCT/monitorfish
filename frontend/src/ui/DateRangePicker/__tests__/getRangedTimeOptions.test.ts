import { expect } from '@jest/globals'

import { getRangedTimeOptions } from '../utils'

describe('ui/DateRangePicker/utils.getRangedTimeOptions()', () => {
  it('should return the expected array of options', () => {
    const minutesRange = 15

    const result = getRangedTimeOptions(minutesRange)

    expect(result).toMatchSnapshot()
  })
})
