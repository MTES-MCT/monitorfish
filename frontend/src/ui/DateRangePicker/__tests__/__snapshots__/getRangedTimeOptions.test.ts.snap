// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/DateRangePicker/utils.getRangedTimeOptions() should return the expected array of options 1`] = `
Array [
  Object {
    "label": "00:00",
    "value": Array [
      0,
      0,
    ],
  },
  Object {
    "label": "00:15",
    "value": Array [
      0,
      15,
    ],
  },
  Object {
    "label": "00:30",
    "value": Array [
      0,
      30,
    ],
  },
  Object {
    "label": "00:45",
    "value": Array [
      0,
      45,
    ],
  },
  Object {
    "label": "01:00",
    "value": Array [
      1,
      0,
    ],
  },
  Object {
    "label": "01:15",
    "value": Array [
      1,
      15,
    ],
  },
  Object {
    "label": "01:30",
    "value": Array [
      1,
      30,
    ],
  },
  Object {
    "label": "01:45",
    "value": Array [
      1,
      45,
    ],
  },
  Object {
    "label": "02:00",
    "value": Array [
      2,
      0,
    ],
  },
  Object {
    "label": "02:15",
    "value": Array [
      2,
      15,
    ],
  },
  Object {
    "label": "02:30",
    "value": Array [
      2,
      30,
    ],
  },
  Object {
    "label": "02:45",
    "value": Array [
      2,
      45,
    ],
  },
  Object {
    "label": "03:00",
    "value": Array [
      3,
      0,
    ],
  },
  Object {
    "label": "03:15",
    "value": Array [
      3,
      15,
    ],
  },
  Object {
    "label": "03:30",
    "value": Array [
      3,
      30,
    ],
  },
  Object {
    "label": "03:45",
    "value": Array [
      3,
      45,
    ],
  },
  Object {
    "label": "04:00",
    "value": Array [
      4,
      0,
    ],
  },
  Object {
    "label": "04:15",
    "value": Array [
      4,
      15,
    ],
  },
  Object {
    "label": "04:30",
    "value": Array [
      4,
      30,
    ],
  },
  Object {
    "label": "04:45",
    "value": Array [
      4,
      45,
    ],
  },
  Object {
    "label": "05:00",
    "value": Array [
      5,
      0,
    ],
  },
  Object {
    "label": "05:15",
    "value": Array [
      5,
      15,
    ],
  },
  Object {
    "label": "05:30",
    "value": Array [
      5,
      30,
    ],
  },
  Object {
    "label": "05:45",
    "value": Array [
      5,
      45,
    ],
  },
  Object {
    "label": "06:00",
    "value": Array [
      6,
      0,
    ],
  },
  Object {
    "label": "06:15",
    "value": Array [
      6,
      15,
    ],
  },
  Object {
    "label": "06:30",
    "value": Array [
      6,
      30,
    ],
  },
  Object {
    "label": "06:45",
    "value": Array [
      6,
      45,
    ],
  },
  Object {
    "label": "07:00",
    "value": Array [
      7,
      0,
    ],
  },
  Object {
    "label": "07:15",
    "value": Array [
      7,
      15,
    ],
  },
  Object {
    "label": "07:30",
    "value": Array [
      7,
      30,
    ],
  },
  Object {
    "label": "07:45",
    "value": Array [
      7,
      45,
    ],
  },
  Object {
    "label": "08:00",
    "value": Array [
      8,
      0,
    ],
  },
  Object {
    "label": "08:15",
    "value": Array [
      8,
      15,
    ],
  },
  Object {
    "label": "08:30",
    "value": Array [
      8,
      30,
    ],
  },
  Object {
    "label": "08:45",
    "value": Array [
      8,
      45,
    ],
  },
  Object {
    "label": "09:00",
    "value": Array [
      9,
      0,
    ],
  },
  Object {
    "label": "09:15",
    "value": Array [
      9,
      15,
    ],
  },
  Object {
    "label": "09:30",
    "value": Array [
      9,
      30,
    ],
  },
  Object {
    "label": "09:45",
    "value": Array [
      9,
      45,
    ],
  },
  Object {
    "label": "10:00",
    "value": Array [
      10,
      0,
    ],
  },
  Object {
    "label": "10:15",
    "value": Array [
      10,
      15,
    ],
  },
  Object {
    "label": "10:30",
    "value": Array [
      10,
      30,
    ],
  },
  Object {
    "label": "10:45",
    "value": Array [
      10,
      45,
    ],
  },
  Object {
    "label": "11:00",
    "value": Array [
      11,
      0,
    ],
  },
  Object {
    "label": "11:15",
    "value": Array [
      11,
      15,
    ],
  },
  Object {
    "label": "11:30",
    "value": Array [
      11,
      30,
    ],
  },
  Object {
    "label": "11:45",
    "value": Array [
      11,
      45,
    ],
  },
  Object {
    "label": "12:00",
    "value": Array [
      12,
      0,
    ],
  },
  Object {
    "label": "12:15",
    "value": Array [
      12,
      15,
    ],
  },
  Object {
    "label": "12:30",
    "value": Array [
      12,
      30,
    ],
  },
  Object {
    "label": "12:45",
    "value": Array [
      12,
      45,
    ],
  },
  Object {
    "label": "13:00",
    "value": Array [
      13,
      0,
    ],
  },
  Object {
    "label": "13:15",
    "value": Array [
      13,
      15,
    ],
  },
  Object {
    "label": "13:30",
    "value": Array [
      13,
      30,
    ],
  },
  Object {
    "label": "13:45",
    "value": Array [
      13,
      45,
    ],
  },
  Object {
    "label": "14:00",
    "value": Array [
      14,
      0,
    ],
  },
  Object {
    "label": "14:15",
    "value": Array [
      14,
      15,
    ],
  },
  Object {
    "label": "14:30",
    "value": Array [
      14,
      30,
    ],
  },
  Object {
    "label": "14:45",
    "value": Array [
      14,
      45,
    ],
  },
  Object {
    "label": "15:00",
    "value": Array [
      15,
      0,
    ],
  },
  Object {
    "label": "15:15",
    "value": Array [
      15,
      15,
    ],
  },
  Object {
    "label": "15:30",
    "value": Array [
      15,
      30,
    ],
  },
  Object {
    "label": "15:45",
    "value": Array [
      15,
      45,
    ],
  },
  Object {
    "label": "16:00",
    "value": Array [
      16,
      0,
    ],
  },
  Object {
    "label": "16:15",
    "value": Array [
      16,
      15,
    ],
  },
  Object {
    "label": "16:30",
    "value": Array [
      16,
      30,
    ],
  },
  Object {
    "label": "16:45",
    "value": Array [
      16,
      45,
    ],
  },
  Object {
    "label": "17:00",
    "value": Array [
      17,
      0,
    ],
  },
  Object {
    "label": "17:15",
    "value": Array [
      17,
      15,
    ],
  },
  Object {
    "label": "17:30",
    "value": Array [
      17,
      30,
    ],
  },
  Object {
    "label": "17:45",
    "value": Array [
      17,
      45,
    ],
  },
  Object {
    "label": "18:00",
    "value": Array [
      18,
      0,
    ],
  },
  Object {
    "label": "18:15",
    "value": Array [
      18,
      15,
    ],
  },
  Object {
    "label": "18:30",
    "value": Array [
      18,
      30,
    ],
  },
  Object {
    "label": "18:45",
    "value": Array [
      18,
      45,
    ],
  },
  Object {
    "label": "19:00",
    "value": Array [
      19,
      0,
    ],
  },
  Object {
    "label": "19:15",
    "value": Array [
      19,
      15,
    ],
  },
  Object {
    "label": "19:30",
    "value": Array [
      19,
      30,
    ],
  },
  Object {
    "label": "19:45",
    "value": Array [
      19,
      45,
    ],
  },
  Object {
    "label": "20:00",
    "value": Array [
      20,
      0,
    ],
  },
  Object {
    "label": "20:15",
    "value": Array [
      20,
      15,
    ],
  },
  Object {
    "label": "20:30",
    "value": Array [
      20,
      30,
    ],
  },
  Object {
    "label": "20:45",
    "value": Array [
      20,
      45,
    ],
  },
  Object {
    "label": "21:00",
    "value": Array [
      21,
      0,
    ],
  },
  Object {
    "label": "21:15",
    "value": Array [
      21,
      15,
    ],
  },
  Object {
    "label": "21:30",
    "value": Array [
      21,
      30,
    ],
  },
  Object {
    "label": "21:45",
    "value": Array [
      21,
      45,
    ],
  },
  Object {
    "label": "22:00",
    "value": Array [
      22,
      0,
    ],
  },
  Object {
    "label": "22:15",
    "value": Array [
      22,
      15,
    ],
  },
  Object {
    "label": "22:30",
    "value": Array [
      22,
      30,
    ],
  },
  Object {
    "label": "22:45",
    "value": Array [
      22,
      45,
    ],
  },
  Object {
    "label": "23:00",
    "value": Array [
      23,
      0,
    ],
  },
  Object {
    "label": "23:15",
    "value": Array [
      23,
      15,
    ],
  },
  Object {
    "label": "23:30",
    "value": Array [
      23,
      30,
    ],
  },
  Object {
    "label": "23:45",
    "value": Array [
      23,
      45,
    ],
  },
]
`;
